### PR TITLE
Sanctions/unsc sanctions xml/comp 1415

### PR DESCRIPTION
**Pull request by saad-deriv**
# Description
sanctions/unsc-sanctions-xml/comp-1415

# Card
* https://app.clickup.com/t/20696747/COMP-1464

### CODE REVIEW

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 1]
```diff
@@ -83,6 +84,11 @@ sub config {
             url         => $eu_url,
             parser      => \&_eu_xml,
         },
+         'UNSC-Sanctions' => {
+            description => 'UN: United Nations Security Council Consolidated List',
+            url         => $args{unsc_url} || 'https://scsanctions.un.org/resources/xml/en/consolidated.xml?_gl=1*17zxk5d*_ga*MTI5MTEzMzMyNS4xNzI5MDY3MjA1*_ga_TK9BQL5X7Z*MTcyOTA2NzIwNC4xLjAuMTcyOTA2NzIwNi4wLjAuMA',
```
Comment (ragheb-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:89]: why are we adding _gl query param?

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 2]
```diff
@@ -425,6 +431,99 @@ sub _eu_xml {
...
+        $entry{dataid}               = $individual->{'DATAID'}[0];
+        $entry{versionnum}           = $individual->{'VERSIONNUM'}[0];
+        $entry{first_name}           = $individual->{'FIRST_NAME'}[0];
+        $entry{second_name}          = $individual->{'SECOND_NAME'}[0];
+        $entry{third_name}           = $individual->{'THIRD_NAME'}[0]  // '';
+        $entry{fourth_name}          = $individual->{'FOURTH_NAME'}[0] // '';
+        $entry{un_list_type}         = $individual->{'UN_LIST_TYPE'}[0];
+        $entry{reference_number}     = $individual->{'REFERENCE_NUMBER'}[0];
+        $entry{listed_on}            = $individual->{'LISTED_ON'}[0];
+        $entry{name_original_script} = $individual->{'NAME_ORIGINAL_SCRIPT'}[0] // '';
```
Comment (ragheb-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:466]: dataid, versionnum, un_list_type, listed_on are not used

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 3]
```diff
@@ -425,6 +431,95 @@ sub _eu_xml {
     };
 }
 
+sub _unsc_xml {
+    my ($xml_content) = @_;
+
+    # Preprocess the XML content to escape unescaped ampersands
+    $xml_content =~ s/&(?!(?:amp|lt|gt|quot|apos);)/&amp;/g;
+    my $data = XML::Simple::XMLin(
```
Comment (ragheb-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:439]: why are we using a parcer here ? 
Can't we use `xml2hash` and reuse the functionalities we had before ? Check the _ofac_xml sub for reference

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 4]
```diff
@@ -425,6 +430,93 @@ sub _eu_xml {
...
+    # Extract the dateGenerated attribute from the first line of the XML content
+    my ($date_generated) = $data->{'-dateGenerated'};
+    die "Corrupt data. Release date is missing\n" unless $date_generated;
+
+    # Convert the dateGenerated to epoch milliseconds
+    my $publish_epoch = _date_to_epoch($date_generated // '');
+
+    my $dataset = [];
+
+    # p $data->{'INDIVIDUALS'}->{'INDIVIDUAL'}->[0];
```
Comment (ragheb-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:449]: is this needed ?

File: Makefile.PL [hunk 5]
```diff
@@ -131,6 +132,7 @@ my %FallbackPrereqs = (
   "Text::CSV" => 0,
   "Text::Trim" => 0,
   "XML::Fast" => 0,
+  "XML::Simple" => 0,
```
Comment (ragheb-deriv) [inline Makefile.PL:135]: this is no longer needed

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 6]
```diff
@@ -425,6 +429,91 @@ sub _eu_xml {
...
+            $entry{name_original_script} // '',
+        );
+
+        foreach my $alias (@{$individual->{INDIVIDUAL_ALIAS}}) {
+            if (ref($alias->{'ALIAS_NAME'}) ne 'HASH' || %{$alias->{'ALIAS_NAME'}}) {
+                push @names, $alias->{'ALIAS_NAME'} // '';
+            }
+        }
+
+        my @dob_list = ($individual->{'INDIVIDUAL_DATE_OF_BIRTH'}[0]{'DATE'} // '');
```
Comment (heshan-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:471]: In some cases the exact date isn't provided. So, `{'INDIVIDUAL_DATE_OF_BIRTH'}[0]{'DATE'}` may not be available. 
In all of the following cases, the date of birth will be an empty string.

`<INDIVIDUAL_DATE_OF_BIRTH>
<TYPE_OF_DATE>EXACT</TYPE_OF_DATE>
<YEAR>1958</YEAR>
</INDIVIDUAL_DATE_OF_BIRTH>`

`<INDIVIDUAL_DATE_OF_BIRTH>
<TYPE_OF_DATE>APPROXIMATELY</TYPE_OF_DATE>
<YEAR>1958</YEAR>
</INDIVIDUAL_DATE_OF_BIRTH>`


`<INDIVIDUAL_DATE_OF_BIRTH>
<TYPE_OF_DATE>BETWEEN</TYPE_OF_DATE>
<FROM_YEAR>1972</FROM_YEAR>
<TO_YEAR>1973</TO_YEAR>
</INDIVIDUAL_DATE_OF_BIRTH>`

Comment (heshan-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:471]: Need to check the array is defined.

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 7]
```diff
@@ -425,6 +429,91 @@ sub _eu_xml {
...
+            $individual->{'INDIVIDUAL_PLACE_OF_BIRTH'}[0]{'STATE_PROVINCE'} // '',
+            $individual->{'INDIVIDUAL_PLACE_OF_BIRTH'}[0]{'COUNTRY'}       // ''
+        );
+
+        my @residence   = (map { $_->{'COUNTRY'}  // '' } @{$individual->{'INDIVIDUAL_ADDRESS'}});
+        my @postal_code = (map { $_->{'ZIP_CODE'} // '' } @{$individual->{'INDIVIDUAL_ADDRESS'}});
+
+        my @nationality = ($individual->{'NATIONALITY'}->{'VALUE'} // '');
+        my @national_id = [];
+        my @passport_no = [];
```
Comment (heshan-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:484]:
```suggestion
        my @national_id;
        my @passport_no;
```

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 8]
```diff
@@ -425,6 +429,91 @@ sub _eu_xml {
...
+        my @names = (
+            $entry{first_name}           // '',
+            $entry{second_name}          // '',
+            $entry{third_name}           // '',
+            $entry{fourth_name}          // '',
+            $entry{name_original_script} // '',
+        );
+
+        foreach my $alias (@{$individual->{INDIVIDUAL_ALIAS}}) {
+            if (ref($alias->{'ALIAS_NAME'}) ne 'HASH' || %{$alias->{'ALIAS_NAME'}}) {
```
Comment (heshan-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:466]: This seems a little off. The condition ref($alias->{'ALIAS_NAME'}) ne 'HASH' is true, then trying to access $alias->{'ALIAS_NAME'} would not be correct.

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 9]
```diff
@@ -425,6 +429,91 @@ sub _eu_xml {
...
+        }
+
+        _process_sanction_entry(
+            $dataset,
+            names          => \@names,
+            date_of_birth  => \@dob_list,
+            place_of_birth => \@place_of_birth,
+            residence      => \@residence,
+            nationality    => \@nationality,
+            citizen        => \@nationality,      # no seprate field for citizenship in the XML
```
Comment (heshan-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:504]: Is it better to leave the citizen field, if it's not defined in the xml?

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 10]
```diff
@@ -425,6 +429,91 @@ sub _eu_xml {
...
+        $entry{second_name}          = $individual->{'SECOND_NAME'};
+        $entry{third_name}           = $individual->{'THIRD_NAME'}           // '';
+        $entry{fourth_name}          = $individual->{'FOURTH_NAME'}          // '';
+        $entry{name_original_script} = $individual->{'NAME_ORIGINAL_SCRIPT'} // '';
+
+        my @names = (
+            $entry{first_name}           // '',
+            $entry{second_name}          // '',
+            $entry{third_name}           // '',
+            $entry{fourth_name}          // '',
```
Comment (heshan-deriv) [inline lib/Data/Validate/Sanctions/Fetcher.pm:461]: I've noticed that in a couple of other places the first_name, second_name and etc, are combined into a single string.
_ofac_xml
_hmt_csv

File: lib/Data/Validate/Sanctions/Fetcher.pm (hunk 6 above)
Comment (saad-deriv) [reply lib/Data/Validate/Sanctions/Fetcher.pm]: done

Comment (saad-deriv) [reply lib/Data/Validate/Sanctions/Fetcher.pm]: done

File: lib/Data/Validate/Sanctions/Fetcher.pm (hunk 10 above)
Comment (saad-deriv) [reply lib/Data/Validate/Sanctions/Fetcher.pm:464]: We can do that but then some individuals have two names present in the first name section e.g. MOHAMMAD NAIM. If we search with MOHAMMAD NAIM we find it in this way, if we join then the array is going to have MOHAMMAD NAIM BARICH instead. Do you think we should have it like that? or Is this way fine?

**Review by heshan-deriv — APPROVED** [review]

**Review by ragheb-deriv — APPROVED** [review]

**Pull request merged**